exercises/src/main/scala/fpinscala/laziness/Stream.scala

```diff
@@ -17,20 +27,95 @@ trait Stream[+A] {
     case Empty => None
     case Cons(h, t) => if (f(h())) Some(h()) else t().find(f)
   }
-  def take(n: Int): Stream[A] = ???
 
-  def drop(n: Int): Stream[A] = ???
+  //5.2
+  def take(n: Int): Stream[A] = this match {
+    case _ if n == 0 => empty
+    case Cons(h, _) if n == 1 => cons(h(), empty)
+    case Cons(h, t) => cons(h(), t().take(n - 1))
+  }
+
+  //5.2
+  def drop(n: Int): Stream[A] = this match {
+    case Cons(_, t) if n >= 1 => t().drop(n - 1)
+    case _ => this
+  }
+
+  //5.3
+  def takeWhile(p: A => Boolean): Stream[A] = this match {
+    case Cons(h, t) if p(h()) => cons(h(), t().takeWhile(p))
+    case _ => empty
+  }
 
-  def takeWhile(p: A => Boolean): Stream[A] = ???
+  //5.4
+  def forAll(p: A => Boolean): Boolean = foldRight(true)((a, b) => p(a) && b )
```

Comment:
With this implementation, if the first predicate fails, would it try to evaluate all of them still?

Reply:
One way to test that, is that if you have an infinite length Stream, then `forAll` should return a value if there's at least one element that fails the predicate. Obviously in that scenario it could never return true though. For example, using Stream from the standard lib:

```
@ def nat(start: Int): Stream[Int] = start #:: nat(start + 1)
defined function nat

@ nat(1).forall(_ % 10 != 0)
res14: Boolean = false

@ nat(1).take(9).forall(_ % 10 != 0)
res15: Boolean = true
```

If you don't want to test crashing your REPL or sbt, you can use a `println` or a variable mutation inside the predicate to test the behavior. Both should be avoided in production code, but that's a legit thing to do for tests reasons.

Reply:
My understanding was that any predicate failing would cause the evaluation to stop.  Becase in foldRight, the second argument 'b' is the recursive call to foldRight, and in ((a, b) => p(a) && b ), b only gets evaluated if p(a) is true

Reply:
hmmm, I'm reading it again, and I'm not sure why I thought it wasn't fail fast, you're right, your implementation appears to be correct. It should be worth having a test that covers it anyway though, to confirm that it works on infinite streams.